### PR TITLE
Fix stuck when using http2 POST Buffer

### LINKED ISSUE #2335: Request gets stuck when doing POST using http2
## Bug Description

When using http2 for POST, if body is string or Buffer, the request will get stuck. This works fine if body is a stream

## Reproducible By

test code:

```ts
test('Should support H2 connection(POST Buffer)', async t => {
  const server = createSecureServer({ ...pem, allowHTTP1: false })

  server.on('stream', async (stream, headers, _flags, rawHeaders) => {
    t.equal(headers[':method'], 'POST')
    const reqData = []
    stream.on('data', chunk => reqData.push(chunk.toString()))
    await once(stream, 'end')
    const reqBody = reqData.join('')
    t.equal(reqBody.length > 0, true)
    stream.respond({
      'content-type': 'text/plain; charset=utf-8',
      'x-custom-h2': 'hello',
      ':status': 200
    })
    stream.end(`hello h2! ${reqBody}`)
  })

  server.listen(0)
  await once(server, 'listening')

  const client = new Client(`https://localhost:${server.address().port}`, {
    connect: {
      rejectUnauthorized: false
    },
    allowH2: true
  })

  t.plan(6)
  t.teardown(server.close.bind(server))
  t.teardown(client.close.bind(client))

  const sendBody = `hello!`
  const body = []
  console.log(`before request ${sendBody}`)
  const response = await client.request({
    path: '/',
    method: 'POST',
    // FIXME: If send a Buffer or string, the stream will not end, causing it to get stuck.
    body: sendBody
  })
  console.log(`after request ${sendBody}`)

  response.body.on('data', chunk => {
    body.push(chunk)
  })

  await once(response.body, 'end')
  t.equal(response.statusCode, 200)
  t.equal(response.headers['content-type'], 'text/plain; charset=utf-8')
  t.equal(response.headers['x-custom-h2'], 'hello')
  t.equal(Buffer.concat(body).toString('utf8'), `hello h2! ${sendBody}`)
})
```

## Logs & Screenshots

image

image

## Environment

- MacOS 14.0
- Nodejs 16.19.0
- undici 5.25.4

### Additional context

Maybe we can fix this by:
- Added call to `stream.end()` in writeBodyH2 ()
- Use `Readable.from(body)` in writeBodyH2 () to convert body to a readable stream

### PR DESCRIPTION
**Pull request by binsee**
## This relates to...

Fixes #2335

### Bug Fixes

- Fix stuck when using http2 POST Buffer

## Status

- [x] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [x] Tested
- [ ] Benchmarked (**optional**)
- [ ] Documented
- [x] Review ready
- [ ] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: test/http2.js [hunk 1]
```diff
@@ -114,6 +114,58 @@ test('Should support H2 connection (headers as array)', async t => {
...
+
+  const sendBody = 'hello!'
+  const body = []
+  console.log(`before request ${sendBody}`)
+  const response = await client.request({
+    path: '/',
+    method: 'POST',
+    body: sendBody
+  })
+  console.log(`after request ${sendBody}`)
```
Comment (metcoder95) [inline test/http2.js:156]: Leftover?

File: test/http2.js [hunk 2]
```diff
@@ -114,6 +114,58 @@ test('Should support H2 connection (headers as array)', async t => {
...
+    allowH2: true
+  })
+
+  t.plan(6)
+  t.teardown(server.close.bind(server))
+  t.teardown(client.close.bind(client))
+
+  const sendBody = 'hello!'
+  const body = []
+  console.log(`before request ${sendBody}`)
```
Comment (metcoder95) [inline test/http2.js:150]: Leftover?

**Review by metcoder95 — COMMENTED** [review]
Nice catch!
Just comments for leftovers

**Comment by binsee** [comment]
@metcoder95 done

**Review by metcoder95 — APPROVED** [review]

**Comment by metcoder95** [comment]
It seems, this closes #2335, is that correct?

**Comment by binsee** [comment]
> It seems, this closes #2335, is that correct?

yes

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**